## 1. The World Bank's international debt data
<p>It's not that we humans only take debts to manage our necessities. A country may also take debt to manage its economy. For example, infrastructure spending is one costly ingredient required for a country's citizens to lead comfortable lives. <a href="https://www.worldbank.org">The World Bank</a> is the organization that provides debt to countries.</p>
<p>In this notebook, we are going to analyze international debt data collected by The World Bank. The dataset contains information about the amount of debt (in USD) owed by developing countries across several categories. We are going to find the answers to questions like: </p>
<ul>
<li>What is the total amount of debt that is owed by the countries listed in the dataset?</li>
<li>Which country owns the maximum amount of debt and what does that amount look like?</li>
<li>What is the average amount of debt owed by countries across different debt indicators?</li>
</ul>
<p><img src="https://assets.datacamp.com/production/project_754/img/image.jpg" alt></p>
<p>The first line of code connects us to the <code>international_debt</code> database where the table <code>international_debt</code> is residing. Let's first <code>SELECT</code> <em>all</em> of the columns from the <code>international_debt</code> table. Also, we'll limit the output to the first ten rows to keep the output clean.</p>

In [4]:
%%sql
postgresql:///international_debt

SELECT * 
FROM international_debt
LIMIT 10
    

10 rows affected.


[('Afghanistan', 'AFG', 'Disbursements on external debt, long-term (DIS, current US$)', 'DT.DIS.DLXF.CD', Decimal('72894453.700000003')),
 ('Afghanistan', 'AFG', 'Interest payments on external debt, long-term (INT, current US$)', 'DT.INT.DLXF.CD', Decimal('53239440.100000001')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (AMT, current US$)', 'DT.AMT.BLAT.CD', Decimal('61739336.899999999')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (DIS, current US$)', 'DT.DIS.BLAT.CD', Decimal('49114729.399999999')),
 ('Afghanistan', 'AFG', 'PPG, bilateral (INT, current US$)', 'DT.INT.BLAT.CD', Decimal('39903620.100000001')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (AMT, current US$)', 'DT.AMT.MLAT.CD', Decimal('39107845')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (DIS, current US$)', 'DT.DIS.MLAT.CD', Decimal('23779724.300000001')),
 ('Afghanistan', 'AFG', 'PPG, multilateral (INT, current US$)', 'DT.INT.MLAT.CD', Decimal('13335820')),
 ('Afghanistan', 'AFG', 'PPG, official creditors (AMT, current US$)', 'DT.AMT.OFFT.CD', Decimal('100847181.900000006')),
 ('Afghanistan', 'AFG', 'PPG, official creditors (DIS, current US$)', 'DT.DIS.OFFT.CD', Decimal('72894453.700000003'))]

## 2. Finding the number of distinct countries
<p>From the first ten rows, we can see the amount of debt owed by <em>Afghanistan</em> in the different debt indicators. But we do not know the number of different countries we have on the table. There are repetitions in the country names because a country is most likely to have debt in more than one debt indicator. </p>
<p>Without a count of unique countries, we will not be able to perform our statistical analyses holistically. In this section, we are going to extract the number of unique countries present in the table. </p>

In [5]:
%%sql
SELECT 
    COUNT(DISTINCT country_name) AS total_distinct_countries
FROM international_debt;

 * postgresql:///international_debt
1 rows affected.


[(124,)]

## 3. Finding out the distinct debt indicators
<p>We can see there are a total of 124 countries present on the table. As we saw in the first section, there is a column called <code>indicator_name</code> that briefly specifies the purpose of taking the debt. Just beside that column, there is another column called <code>indicator_code</code> which symbolizes the category of these debts. Knowing about these various debt indicators will help us to understand the areas in which a country can possibly be indebted to. </p>

In [6]:
%%sql
SELECT DISTINCT indicator_code AS distinct_debt_indicators
FROM international_debt
ORDER BY distinct_debt_indicators

 * postgresql:///international_debt
25 rows affected.


[('DT.AMT.BLAT.CD',),
 ('DT.AMT.DLXF.CD',),
 ('DT.AMT.DPNG.CD',),
 ('DT.AMT.MLAT.CD',),
 ('DT.AMT.OFFT.CD',),
 ('DT.AMT.PBND.CD',),
 ('DT.AMT.PCBK.CD',),
 ('DT.AMT.PROP.CD',),
 ('DT.AMT.PRVT.CD',),
 ('DT.DIS.BLAT.CD',),
 ('DT.DIS.DLXF.CD',),
 ('DT.DIS.MLAT.CD',),
 ('DT.DIS.OFFT.CD',),
 ('DT.DIS.PCBK.CD',),
 ('DT.DIS.PROP.CD',),
 ('DT.DIS.PRVT.CD',),
 ('DT.INT.BLAT.CD',),
 ('DT.INT.DLXF.CD',),
 ('DT.INT.DPNG.CD',),
 ('DT.INT.MLAT.CD',),
 ('DT.INT.OFFT.CD',),
 ('DT.INT.PBND.CD',),
 ('DT.INT.PCBK.CD',),
 ('DT.INT.PROP.CD',),
 ('DT.INT.PRVT.CD',)]

## 4. Totaling the amount of debt owed by the countries
<p>As mentioned earlier, the financial debt of a particular country represents its economic state. But if we were to project this on an overall global scale, how will we approach it?</p>
<p>Let's switch gears from the debt indicators now and find out the total amount of debt (in USD) that is owed by the different countries. This will give us a sense of how the overall economy of the entire world is holding up.</p>

In [9]:
%%sql
SELECT 
    ROUND(debt / 1000000, 2) AS total_debt
FROM international_debt; 

 * postgresql:///international_debt
2357 rows affected.


[(Decimal('72.89'),),
 (Decimal('53.24'),),
 (Decimal('61.74'),),
 (Decimal('49.11'),),
 (Decimal('39.90'),),
 (Decimal('39.11'),),
 (Decimal('23.78'),),
 (Decimal('13.34'),),
 (Decimal('100.85'),),
 (Decimal('72.89'),),
 (Decimal('53.24'),),
 (Decimal('100.85'),),
 (Decimal('317.19'),),
 (Decimal('165.60'),),
 (Decimal('87.88'),),
 (Decimal('54.25'),),
 (Decimal('76.05'),),
 (Decimal('13.85'),),
 (Decimal('0.00'),),
 (Decimal('31.03'),),
 (Decimal('39.45'),),
 (Decimal('4.54'),),
 (Decimal('4.62'),),
 (Decimal('182.20'),),
 (Decimal('234.32'),),
 (Decimal('28.10'),),
 (Decimal('236.45'),),
 (Decimal('310.37'),),
 (Decimal('41.95'),),
 (Decimal('0.17'),),
 (Decimal('2.28'),),
 (Decimal('0.12'),),
 (Decimal('39.62'),),
 (Decimal('6.82'),),
 (Decimal('35.77'),),
 (Decimal('790.25'),),
 (Decimal('514.19'),),
 (Decimal('10.32'),),
 (Decimal('19.03'),),
 (Decimal('3.22'),),
 (Decimal('95.19'),),
 (Decimal('10.32'),),
 (Decimal('7.68'),),
 (Decimal('0.58'),),
 (Decimal('0.01'),),
 (Decimal('0.00'),),
 (Decimal('8.09'),),
 (Decimal('95.19'),),
 (Decimal('10.32'),),
 (Decimal('15.78'),),
 (Decimal('0.00'),),
 (Decimal('0.02'),),
 (Decimal('0.58'),),
 (Decimal('0.04'),),
 (Decimal('171.19'),),
 (Decimal('75.42'),),
 (Decimal('10924.02'),),
 (Decimal('1798.55'),),
 (Decimal('8473.82'),),
 (Decimal('8838.26'),),
 (Decimal('1005.05'),),
 (Decimal('1000.00'),),
 (Decimal('70.00'),),
 (Decimal('1125.09'),),
 (Decimal('1906.77'),),
 (Decimal('580.90'),),
 (Decimal('124.69'),),
 (Decimal('178.99'),),
 (Decimal('47.04'),),
 (Decimal('8598.51'),),
 (Decimal('9017.25'),),
 (Decimal('1052.09'),),
 (Decimal('343.44'),),
 (Decimal('95.56'),),
 (Decimal('2468.53'),),
 (Decimal('1906.77'),),
 (Decimal('746.46'),),
 (Decimal('11067.05'),),
 (Decimal('426.96'),),
 (Decimal('275.29'),),
 (Decimal('128.08'),),
 (Decimal('68.97'),),
 (Decimal('174.27'),),
 (Decimal('24.09'),),
 (Decimal('0.00'),),
 (Decimal('65.75'),),
 (Decimal('1.64'),),
 (Decimal('0.20'),),
 (Decimal('94.33'),),
 (Decimal('252.69'),),
 (Decimal('57.17'),),
 (Decimal('163.30'),),
 (Decimal('426.96'),),
 (Decimal('81.27'),),
 (Decimal('1.64'),),
 (Decimal('65.95'),),
 (Decimal('845.63'),),
 (Decimal('680.70'),),
 (Decimal('1088.46'),),
 (Decimal('495.73'),),
 (Decimal('74.18'),),
 (Decimal('106.37'),),
 (Decimal('84.75'),),
 (Decimal('10.09'),),
 (Decimal('0.00'),),
 (Decimal('225.02'),),
 (Decimal('423.61'),),
 (Decimal('347.55'),),
 (Decimal('49.57'),),
 (Decimal('513.62'),),
 (Decimal('656.16'),),
 (Decimal('136.58'),),
 (Decimal('619.99'),),
 (Decimal('740.91'),),
 (Decimal('146.66'),),
 (Decimal('4.11'),),
 (Decimal('0.29'),),
 (Decimal('427.72'),),
 (Decimal('347.55'),),
 (Decimal('274.89'),),
 (Decimal('1513.81'),),
 (Decimal('466.10'),),
 (Decimal('9050.56'),),
 (Decimal('821.15'),),
 (Decimal('155.54'),),
 (Decimal('386.70'),),
 (Decimal('6141.79'),),
 (Decimal('309.08'),),
 (Decimal('3.94'),),
 (Decimal('0.00'),),
 (Decimal('1039.56'),),
 (Decimal('2908.77'),),
 (Decimal('356.39'),),
 (Decimal('1426.27'),),
 (Decimal('9050.56'),),
 (Decimal('665.47'),),
 (Decimal('1.68'),),
 (Decimal('0.13'),),
 (Decimal('5.63'),),
 (Decimal('0.13'),),
 (Decimal('2077.01'),),
 (Decimal('645.12'),),
 (Decimal('2525.23'),),
 (Decimal('1214.49'),),
 (Decimal('272.64'),),
 (Decimal('1487.36'),),
 (Decimal('2265.34'),),
 (Decimal('630.23'),),
 (Decimal('0.00'),),
 (Decimal('100.75'),),
 (Decimal('935.06'),),
 (Decimal('77.10'),),
 (Decimal('595.44'),),
 (Decimal('241.49'),),
 (Decimal('133.03'),),
 (Decimal('2082.80'),),
 (Decimal('2506.83'),),
 (Decimal('763.26'),),
 (Decimal('28.91'),),
 (Decimal('18.40'),),
 (Decimal('0.74'),),
 (Decimal('963.97'),),
 (Decimal('18.40'),),
 (Decimal('178.59'),),
 (Decimal('4640.04'),),
 (Decimal('1593.27'),),
 (Decimal('53.43'),),
 (Decimal('53.57'),),
 (Decimal('4.48'),),
 (Decimal('23.33'),),
 (Decimal('3.58'),),
 (Decimal('3.46'),),
 (Decimal('26.33'),),
 (Decimal('33.55'),),
 (Decimal('0.00'),),
 (Decimal('1.08'),),
 (Decimal('29.68'),),
 (Decimal('49.85'),),


## 5. Country with the highest debt
<p>"Human beings cannot comprehend very large or very small numbers. It would be useful for us to acknowledge that fact." - <a href="https://en.wikipedia.org/wiki/Daniel_Kahneman">Daniel Kahneman</a>. That is more than <em>3 million <strong>million</strong></em> USD, an amount which is really hard for us to fathom. </p>
<p>Now that we have the exact total of the amounts of debt owed by several countries, let's now find out the country that owns the highest amount of debt along with the amount. <strong>Note</strong> that this debt is the sum of different debts owed by a country across several categories. This will help to understand more about the country in terms of its socio-economic scenarios. We can also find out the category in which the country owns its highest debt. But we will leave that for now. </p>

In [12]:
%%sql
SELECT 
    country_name, 
    SUM(debt) AS total_debt
FROM international_debt
GROUP BY country_name
ORDER BY total_debt DESC
LIMIT 1;

 * postgresql:///international_debt
1 rows affected.


[('China', Decimal('285793494734.200001568'))]

## 6. Average amount of debt across indicators
<p>So, it was <em>China</em>. A more in-depth breakdown of China's debts can be found <a href="https://datatopics.worldbank.org/debt/ids/country/CHN">here</a>. </p>
<p>We now have a brief overview of the dataset and a few of its summary statistics. We already have an idea of the different debt indicators in which the countries owe their debts. We can dig even further to find out on an average how much debt a country owes? This will give us a better sense of the distribution of the amount of debt across different indicators.</p>

In [16]:
%%sql
SELECT 
    indicator_code AS debt_indicator,
    indicator_name,
    AVG(debt) AS average_debt
FROM international_debt
GROUP BY debt_indicator, indicator_name
ORDER BY average_debt DESC
LIMIT 10;

 * postgresql:///international_debt
10 rows affected.


[('DT.AMT.DLXF.CD', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('5904868401.499193612')),
 ('DT.AMT.DPNG.CD', 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', Decimal('5161194333.812658349')),
 ('DT.DIS.DLXF.CD', 'Disbursements on external debt, long-term (DIS, current US$)', Decimal('2152041216.890243888')),
 ('DT.DIS.OFFT.CD', 'PPG, official creditors (DIS, current US$)', Decimal('1958983452.859836046')),
 ('DT.AMT.PRVT.CD', 'PPG, private creditors (AMT, current US$)', Decimal('1803694101.963265321')),
 ('DT.INT.DLXF.CD', 'Interest payments on external debt, long-term (INT, current US$)', Decimal('1644024067.650806481')),
 ('DT.DIS.BLAT.CD', 'PPG, bilateral (DIS, current US$)', Decimal('1223139290.398230108')),
 ('DT.INT.DPNG.CD', 'Interest payments on external debt, private nonguaranteed (PNG) (INT, current US$)', Decimal('1220410844.421518983')),
 ('DT.AMT.OFFT.CD', 'PPG, official creditors (AMT, current US$)', Decimal('1191187963.083064523')),
 ('DT.AMT.PBND.CD', 'PPG, bonds (AMT, current US$)', Decimal('1082623947.653623188'))]

## 7. The highest amount of principal repayments
<p>We can see that the indicator <code>DT.AMT.DLXF.CD</code> tops the chart of average debt. This category includes repayment of long term debts. Countries take on long-term debt to acquire immediate capital. More information about this category can be found <a href="https://datacatalog.worldbank.org/principal-repayments-external-debt-long-term-amt-current-us-0">here</a>. </p>
<p>An interesting observation in the above finding is that there is a huge difference in the amounts of the indicators after the second one. This indicates that the first two indicators might be the most severe categories in which the countries owe their debts.</p>
<p>We can investigate this a bit more so as to find out which country owes the highest amount of debt in the category of long term debts (<code>DT.AMT.DLXF.CD</code>). Since not all the countries suffer from the same kind of economic disturbances, this finding will allow us to understand that particular country's economic condition a bit more specifically. </p>

In [25]:
%%sql
SELECT 
    country_name, 
    indicator_name
FROM international_debt
WHERE debt = (SELECT 
                 MAX(debt)
             FROM international_debt
             WHERE indicator_code = 'DT.AMT.DLXF.CD');

 * postgresql:///international_debt
1 rows affected.


[('China', 'Principal repayments on external debt, long-term (AMT, current US$)')]

## 8. The most common debt indicator
<p>China has the highest amount of debt in the long-term debt (<code>DT.AMT.DLXF.CD</code>) category. This is verified by <a href="https://data.worldbank.org/indicator/DT.AMT.DLXF.CD?end=2018&most_recent_value_desc=true">The World Bank</a>. It is often a good idea to verify our analyses like this since it validates that our investigations are correct. </p>
<p>We saw that long-term debt is the topmost category when it comes to the average amount of debt. But is it the most common indicator in which the countries owe their debt? Let's find that out. </p>

In [26]:
%%sql
SELECT indicator_code, COUNT(indicator_code) AS indicator_count
FROM international_debt
GROUP BY indicator_code
ORDER BY indicator_count DESC, indicator_code DESC
LIMIT 20


 * postgresql:///international_debt
20 rows affected.


[('DT.INT.OFFT.CD', 124),
 ('DT.INT.MLAT.CD', 124),
 ('DT.INT.DLXF.CD', 124),
 ('DT.AMT.OFFT.CD', 124),
 ('DT.AMT.MLAT.CD', 124),
 ('DT.AMT.DLXF.CD', 124),
 ('DT.DIS.DLXF.CD', 123),
 ('DT.INT.BLAT.CD', 122),
 ('DT.DIS.OFFT.CD', 122),
 ('DT.AMT.BLAT.CD', 122),
 ('DT.DIS.MLAT.CD', 120),
 ('DT.DIS.BLAT.CD', 113),
 ('DT.INT.PRVT.CD', 98),
 ('DT.AMT.PRVT.CD', 98),
 ('DT.INT.PCBK.CD', 84),
 ('DT.AMT.PCBK.CD', 84),
 ('DT.INT.DPNG.CD', 79),
 ('DT.AMT.DPNG.CD', 79),
 ('DT.INT.PBND.CD', 69),
 ('DT.AMT.PBND.CD', 69)]

## 9. Other viable debt issues and conclusion
<p>There are a total of six debt indicators in which all the countries listed in our dataset have taken debt. The indicator <code>DT.AMT.DLXF.CD</code> is also there in the list. So, this gives us a clue that all these countries are suffering from a common economic issue. But that is not the end of the story, a part of the story rather. </p>
<p>Let's change tracks from <code>debt_indicator</code>s now and focus on the amount of debt again. Let's find out the maximum amount of debt across the indicators along with the respective country names. With this, we will be in a position to identify the other plausible economic issues a country might be going through. By the end of this section, we will have found out the debt indicators in which a country owes its highest debt. </p>
<p>In this notebook, we took a look at debt owed by countries across the globe. We extracted a few summary statistics from the data and unraveled some interesting facts and figures. We also validated our findings to make sure the investigations are correct.</p>

In [27]:
%%sql
SELECT country_name, indicator_code, MAX(debt) AS maximum debt
FROM international_debt
GROUP BY country_name, indicator_code
ORDER BY maxium_debt DESC
LIMIT 10

 * postgresql:///international_debt
(psycopg2.ProgrammingError) syntax error at or near "debt"
LINE 1: ...LECT country_name, indicator_code, MAX(debt) AS maximum debt
                                                                   ^
 [SQL: 'SELECT country_name, indicator_code, MAX(debt) AS maximum debt\nFROM international_debt\nGROUP BY country_name, indicator_code\nORDER BY maxium_debt DESC\nLIMIT 10'] (Background on this error at: http://sqlalche.me/e/f405)
